File: codechef/AugustLongChallenge/zombies.cpp
#include <bits/stdc++.h> 

using namespace std; 

int main() {
    int T; 
    cin >> T; 

    for (int t = 0; t < T; t++) {
        int n; 
        cin >> n; 

        vector<unsigned int> radiations(n), health(n), result(n);
        //cout<<"\n"<<radiations.size() << " " << health.size()<<"\n";
        for (int i = 0; i < n; i++) {
                cin >> radiations[i];  
        }
       
        unsigned int tmp;
        for (int i = 0; i < n; i++) {
            //cout<<"\nhealth i "<<i;
            //cin>>tmp;
            //health[i] = tmp;
            cin >> health[i];
            //cout<<"\ngot health "<<i;
        }

        //cout<<"got input"; 
        unsigned int currentRad = 0, temp;
        long int extent;
        map<unsigned int, int> healthMap; 
        for (int i = 0; i < n; i++) {
            temp = currentRad; 

            extent = i + radiations[i]; 
            if (extent < n) {
                result[extent] += 1; 
            }

            currentRad = currentRad - result[i] + 1; 
            result[i] = temp; 
            
            
            map<unsigned int, int>::iterator it = healthMap.find(health[i]); 
            if (it != healthMap.end()) {
                healthMap[health[i]] += 1;  
            } else {
                healthMap[health[i]] = 1;
            }
        }

        currentRad = 0; 
        unsigned int matches = 0; 
        vector<unsigned int> debts(n, 0);
        
        for (int i = n-1; i >= 0; i--) {
            result[i] += currentRad + 1; 

            extent = i - radiations[i]; 
            if (extent > 0) {
                debts[extent] += 1; 
            }

            currentRad = currentRad - debts[i] + 1;  
            map<unsigned int, int>::iterator it = healthMap.find(health[i]); 
            if (it != healthMap.end()) {
                if (healthMap[result[i]] < 1) {
                    break; 
                } else {
                    healthMap[result[i]] -= 1;
                    matches += 1; 
                }
            } else {
                break; 
            }
        }
    
        if (matches == n) {
            cout <<"YES"<<"\n";
        } else {
            cout <<"NO"<<"\n";
        }
    }
    
    return 0; 
}

File: codechef/AugustLongChallenge/zombies.py
T = int(input())
yes = "YES"
no = "NO"

for t in range(T): 
    n = int(input())
    radiations = list(map(int, input().split()))
    health = list(map(int, input().split()))
    healthMap = dict() 
    result = [0] * n

    currentRad = 0
    for i in range(len(radiations)):
        temp = currentRad 

        extent = i + radiations[i]
        if extent < n: 
            result[extent] += 1 

        currentRad = currentRad - result[i] + 1 
        result[i] = temp 

        if health[i] in healthMap: 
            healthMap[health[i]] += 1
        else: 
            healthMap[health[i]] = 1

    currentRad = 0
    debts = [0] * n 
    matches = 0

    for i in range(len(radiations)-1, -1, -1):
        result[i] += currentRad + 1

        extent = i - radiations[i]
        if extent > 0:
            debts[extent] += 1

        currentRad = currentRad - debts[i] + 1
        if result[i] in healthMap: 
            count = healthMap[result[i]]
            if count < 1: 
                break 
            else: 
                healthMap[result[i]] -= 1
                matches += 1
        else: 
            break

    if matches == n: 
        print(yes)
    else: 
        print(no)
    

